### PR TITLE
Implement eval-able AnnotatedString 2-arg show

### LINKED ISSUE #54193: `AnnotatedString`s are not repr-eval-able.
`eval(Meta.parse(repr(x))) == x` should hold whenever possible. It does not hold for `AnnotatedString`s.

An example of a symptom of this problem that came up when testing #53715 is that `@test` equality hints are misleading for `AnotatedString`s with annotation mismatches:

Screenshot of MWE with color
```julia
julia> using StyledStrings, Test

julia> x = styled"{red:abc}"
"abc"

julia> @test x == "abc"
Test Failed at REPL[4]:1
  Expression: x == "abc"
   Evaluated: "abc" == "abc"

ERROR: There was an error during testing

julia> repr(x)
"\"abc\""
```

Adding a special case that notes "annotation mismatch" would be an improvement, but I'm guessing `repr` is the root cause.

cc @tecosaur

### PR DESCRIPTION
**Pull request by tecosaur**
Resolves #54193, which see.

More details in the commit message :slightly_smiling_face:

### CODE REVIEW

**Comment by LilithHafner** [comment]
Should also define the 3-arg method to keep using the AbstractString method

```julia
julia> @eval Base function show(io::IO, s::A) where {A <: AnnotatedString}
           show(io, A)
           print(io, '(')
           show(io, s.string)
           print(io, ", ")
           show(IOContext(io, :typeinfo => typeof(annotations(s))), annotations(s))
           print(io, ')')
       end
show (generic function with 319 methods)

julia> using StyledStrings

julia> styled"{red:hi}"
Base.AnnotatedString{String}("hi", [(1:2, :face => :red)])

julia> @eval Base show(io::IO, ::MIME"text/plain", s::A) where {A <: AnnotatedString} = invoke(show, Tuple{IO, AbstractString}, io, s)
show (generic function with 320 methods)

julia> styled"{red:hi}"
"hi" # this is red

julia> @test ans == "hi"
Test Failed at REPL[13]:1
  Expression: ans == "hi"
   Evaluated: Base.AnnotatedString{String}("hi", [(1:2, :face => :red)]) == "hi"

ERROR: There was an error during testing
```

**Comment by tecosaur** [comment]
Ah yep, thanks for catching that.

File: test/strings/annotated.jl [hunk 1]
```diff
@@ -58,6 +58,9 @@
                              (3:3, :val => 2)])),
              Base.AnnotatedString("abc", [(1:2, :val => 1),
                                   (2:3, :val => 2)]))
+    @test chopprefix(sprint(show, str), "Base.") ==
+        "AnnotatedString{String}(\"some string\", , [(1:4, :thing=>0x01), (6:11, :other=>0x02), (1:11, :all=>0x03)])"
+    @test sprint(show, str, MIME("text/plain")) == "\"some string\""
```
Comment (LilithHafner) [inline test/strings/annotated.jl:63]:
```suggestion
    @test sprint(show, str, MIME("text/plain")) == "\"some string\""
    @test eval(Meta.parse(repr(str))) == str
```
Just to make sure it repr-evals.

**Review by LilithHafner — APPROVED** [review]

**Comment by tecosaur** [comment]
Thanks for giving this a look so quickly :heart_eyes:

**Comment by tecosaur** [comment]
Ah, I see the test introduced here will have to be adjusted in #54289.

**Pull request merged**